torch/utils/mem_tracker.py

```diff
@@ -0,0 +1,626 @@
+        update_existing: bool = False,
+    ) -> Self:
+        st = cls.get_untyped_storage(t)
+        if winfo := cls.get_winfo_if_exists(st, WINFO):
+            winfo.update_reftype(reftype)
+            return cast(cls, winfo)
+        elif update_existing:
+            raise ValueError("No existing winfo found")
+        else:
+            winfo = cls.create_winfo(st, reftype, WINFO)
+            return cast(cls, winfo)
+
+
+class MemTracker(TorchDispatchMode):
+    """
+    MemTracker is a TorchDispatchMode that tracks the memory usage of the
+    parameters, buffers, activations, gradients and optimizer states within its
+    context. It captures "memory snapshots" of the memory usage of the model at
+    various points in time (Pre-Forward, Post-Forward, Pre-Backward and Post-
+    Backward).
+
+    Optional kwargs:
+
+    depth (int): The depth of the module hierarchy to display,
+                defaults to 2.
+
+    units (str): The units to use for memory tracking, defaults to "MB",
+                supports ["B" "KB", "MB", "GB"].
+
+    display_modulewise_stats (bool): Whether to display modulewise memory stats,
+                                     defaults to False.
+
+    display_peak_stats (bool): Whether to display the final peak memory stats,
+                            defaults to True.
+
+    Methods:
+
+    track_external(*obj): Track the memory usage of the objects instantiated
+                        outside the MemTracker.
+
+    print_mem_snapshot(): Captures and prints the current memory snapshot.
+
+    get_peak_mem(): Returns the peak memory seen so far.
+
+    print_peak_mem_snapshot(): Returns the memory breakdown snapshot at the peak
+                            memory seen so far.
+
+    Contract:
+
+    The MemTracker should be used as a context manager.
+    The modules, optimizers and any other tensors created within the context of
+    MemTracker will be tracked by default.
+
+    All the modules, optimizers and tensors that need to be tracked but are
+    created outside the MemTracker should be registered using the
+    track_external method. The track_external method should be called before
+    the MemTracker is used. Any tensors that are supplied to the
+    track_external method that are not module parameters, buffers, gradients
+    activations or optimizer states will be catergorized as "Other".
+    If you want them categorized with a custom name, please file a github issue.
+
+    Any tensors created outside the MemTracker and not supplied to the
+    track_external method will be tracked only after their first use inside
+    the context of MemTracker.
+
+
+        Example usage:
+        .. code-block:: python
+
+            with MemTracker() as mt:
+                module = ...
+                optimizer = ...
+                inp = ...
+                optimizer.zero_grad()
+                loss = module(inp)
+                loss.backward()
+                optimizer.step()
```

Comment:
Should these examples show how to extract the result for the MemTracker?

Reply:
I have so the display module wise stats is a way to look at the output. But do you want me to provide an API to access internal structures as well?

Reply:
Ho you automatically print on exit of the context manager. Sorry didn't see the code. Do we do that for the flop_counter? That sounds surprising to me.

Reply:
Yes, there are two options in the init function, display_module_wise_sats and display_peak_stats. And yes we show them at the exit of the context manager.